Sources/NIO/SelectorUring.swift

```diff
@@ -0,0 +1,353 @@
+                                    sequenceIdentifier: sequenceIdentifier,
+                                    submitNow: !deferReregistrations,
+                                    multishot: multishot)
+    }
+
+    func reregister0<S: Selectable>(selectable: S, fd: Int, oldInterested: SelectorEventSet, newInterested: SelectorEventSet, sequenceIdentifier: RegistrationSequenceIdentifier) throws {
+        _debugPrint("Re-register old \(oldInterested) new \(newInterested) uringEventSet [\(oldInterested.uringEventSet)] reg.uringEventSet [\(newInterested.uringEventSet)]")
+
+        deferredReregistrationsPending = true
+        if multishot {
+            ring.io_uring_poll_update(fd: Int32(fd),
+                                      newPollmask: newInterested.uringEventSet,
+                                      oldPollmask: oldInterested.uringEventSet,
+                                      sequenceIdentifier: sequenceIdentifier,
+                                      submitNow: !deferReregistrations,
+                                      multishot: true)
+        } else {
+            ring.io_uring_prep_poll_remove(fd: Int32(fd),
+                                           pollMask: oldInterested.uringEventSet,
+                                           sequenceIdentifier: sequenceIdentifier,
+                                           submitNow:!deferReregistrations,
+                                           link: true) // next event linked will cancel if this event fails
+
+            ring.io_uring_prep_poll_add(fd: Int32(fd),
+                                        pollMask: newInterested.uringEventSet,
+                                        sequenceIdentifier: sequenceIdentifier,
+                                        submitNow: !deferReregistrations,
+                                        multishot: false)
+        }
+    }
+
+    func deregister0<S: Selectable>(selectable: S, fd: Int, oldInterested: SelectorEventSet, sequenceIdentifier: RegistrationSequenceIdentifier) throws {
+        _debugPrint("deregister interested \(selectable) reg.interested.uringEventSet [\(oldInterested.uringEventSet)]")
+
+        deferredReregistrationsPending = true
+        ring.io_uring_prep_poll_remove(fd: Int32(fd),
+                                       pollMask: oldInterested.uringEventSet,
+                                       sequenceIdentifier: sequenceIdentifier,
+                                       submitNow:!deferReregistrations)
+    }
+
+    private func _shouldRefreshPollForEvent(selectorEvent:SelectorEventSet) -> Bool {
+        if selectorEvent.contains(.read) {
+            // as we don't do exhaustive reads, we need to prod the kernel for
+            // new events, would be even better if we knew if we had read all there is
+            return true
+        }
+        // FIXME: Question for SwiftNIO team: Need to verify that SwiftNIO writes until blocking (exhaustive writing)
```

Comment:
Yes, at the moment, NIO (with `BaseSocketChannel`) will always write until we get `EWOULDBLOCK` (or of course nothing more to write / error / ...).